pageserver/src/tenant/storage_layer.rs

```diff
@@ -127,14 +108,40 @@ pub trait Layer: Send + Sync {
         reconstruct_data: &mut ValueReconstructState,
     ) -> Result<ValueReconstructResult>;
 
-    /// Does this layer only contain some data for the key-range (incremental),
-    /// or does it contain a version of every page? This is important to know
-    /// for garbage collecting old layers: an incremental layer depends on
-    /// the previous non-incremental layer.
-    fn is_incremental(&self) -> bool;
+    /// A short ID string that uniquely identifies the given layer within a [`LayerMap`].
+    fn short_id(&self) -> String;
+
+    /// Dump summary of the contents of the layer to stdout
+    fn dump(&self, verbose: bool) -> Result<()>;
+}
+
+/// A Layer contains all data in a "rectangle" consisting of a range of keys and
+/// range of LSNs.
+///
+/// There are two kinds of layers, in-memory and on-disk layers. In-memory
+/// layers are used to ingest incoming WAL, and provide fast access to the
+/// recent page versions. On-disk layers are stored as files on disk, and are
+/// immutable. This trait presents the common functionality of in-memory and
+/// on-disk layers.
+///
+/// Furthermore, there are two kinds of on-disk layers: delta and image layers.
+/// A delta layer contains all modifications within a range of LSNs and keys.
+/// An image layer is a snapshot of all the data in a key-range, at a single
+/// LSN
+///
+pub trait Layer: Send + Sync + PureLayer {
+    fn get_tenant_id(&self) -> TenantId;
+
+    /// Identify the timeline this layer belongs to
+    fn get_timeline_id(&self) -> TimelineId;
+
+    /// Filename used to store this layer on disk. (Even in-memory layers
+    /// implement this, to print a handy unique identifier for the layer for
+    /// log messages, even though they're never not on disk.)
+    fn filename(&self) -> LayerFileName;
 
-    /// Returns true for layers that are represented in memory.
-    fn is_in_memory(&self) -> bool;
+    /// If a layer has a corresponding file on a local filesystem, return its absolute path.
+    fn local_path(&self) -> Option<LayerFileName>;
```

Comment:
In every current implementation, this is `Some`.

I think, the trait naming is odd: the key feature of the `Layer` now after the refactoring seems to be file persistence.
We have a `HistoricLayer` term for that, we might use that or rename that term alltogether.

Not sure what "Pure" layers represent due to missing rustdoc, but I'd rather call them "stateless"?
Maybe just name _that_ trait `Layer` instead?


---

Let's also adjust `key_iter` and remove the default implementation, if you agree with the concept of the split.

Reply:
> In every current implementation, this is `Some`.

All current implementations return `Some`, but with on-demand download,
a new `RemoteLayer` type will be added that returns `None` here.
No sense in churning so much, IMO.
Will add a comment.

Reply:
> 
> I think, the trait naming is odd: the key feature of the `Layer` now after the refactoring seems to be file persistence.
We have a `HistoricLayer` term for that, we might use that or rename that term alltogether.

Was thinking along the same lines, but didn't want to cause more churn in areas of the code that don't use `dyn Layer`.
Maybe a third commit that completes the renaming?